### PR TITLE
Make more operators support `torch.complex32`

### PR DESCRIPTION
**Pull request by malfet**
Collection of changes aimed at allowing developer to create complex32 tensor,print it and access its elements
A followup after https://github.com/pytorch/pytorch/pull/43279

### CODE REVIEW

File: torch/csrc/utils/python_scalars.h [hunk 1]
```diff
@@ -40,8 +41,12 @@ inline PyObject* load_scalar(void* data, at::ScalarType scalarType) {
     case at::kHalf: return PyFloat_FromDouble(at::convert<double, at::Half>(*(at::Half*)data));
     case at::kFloat: return PyFloat_FromDouble(*(float*)data);
     case at::kDouble: return PyFloat_FromDouble(*(double*)data);
+    case at::kComplexHalf: {
+      auto data_ = (c10::complex<at::Half>*)data;
```
Comment (anjali411) [inline torch/csrc/utils/python_scalars.h:45]: `auto data_ = reinterpret_cast<c10::complex<at::Half>*>(data);`

File: c10/util/Half.h [hunk 2]
```diff
@@ -374,6 +374,8 @@ struct alignas(4) complex<Half> {
   }
   inline complex(c10::complex<float> value)
       : real_(value.real()), imag_(value.imag()) {}
+  inline complex(c10::complex<double> value)
```
Comment (anjali411) [inline c10/util/Half.h:377]: what is this change for?

**Review by anjali411 — CHANGES_REQUESTED** [review]
please add tests to verify correctness of `torch.masked_select`, `torch.masked_fill` for BFloat16  and `torch.lt` for `Half`.

File: c10/util/Half.h (hunk 2 above)
Comment (malfet) [reply c10/util/Half.h:377]: To support `torch::utils::store_scalar()` for c10::complex<Half>

**Comment by malfet** [comment]
@anjali411 do you want an explicit or implicit test? `print(torch.tensor([1+2j], dtype=torch.complex32))` implicitly tests all those operators. And can you please point me to the test that validates torch.lt for all dtypes?

File: torch/testing/__init__.py [hunk 3]
```diff
@@ -313,20 +313,25 @@ def all_types_and_complex_and(*dtypes):
...
+        dtypes += get_all_complex_dtypes(include_complex32)
     return dtypes
 
 def get_all_math_dtypes(device) -> List[torch.dtype]:
     return get_all_int_dtypes() + get_all_fp_dtypes(include_half=device.startswith('cuda'),
                                                     include_bfloat16=False) + get_all_complex_dtypes()
 
-def get_all_complex_dtypes() -> List[torch.dtype]:
-    return [torch.complex64, torch.complex128]
+def get_all_complex_dtypes(include_complex32=False) -> List[torch.dtype]:
```
Comment (anjali411) [inline torch/testing/__init__.py:333]: nit - `include_chalf=False` for a more consistent naming scheme. Also we should add the alias `chalf` for `complex32`

File: aten/src/ATen/cpu/vec256/vec256_base.h [hunk 4]
```diff
@@ -312,10 +312,10 @@ struct Vec256 {
     return ret;
   }
   Vec256<T> erf() const {
-    return map(std::erf);
+    return map([](T x) -> T { return std::erf(x); });
```
Comment (anjali411) [inline aten/src/ATen/cpu/vec256/vec256_base.h:315]: what is this and other similar changes for?

**Review by anjali411 — CHANGES_REQUESTED** [review]
Currently this PR has some missing tests, for example, `torch.lt`, `torch.sqrt`, `index_put` etc. Please add tests for all ops enabled for `torch.half` and `torch.complex32`.

Also, I think we should add the alias `torch.chalf` for `torch.complex32` similar to numpy aliases `np.cfloat` for `np.complex64` and `np.cdouble` for np.complex128`. This can be done in a separate PR, but just added this note as an FYI.

Also we could add a section ComplexHalf support [here](https://github.com/pytorch/pytorch/issues/33152) to keep track of all the support being added and action items on the list. Would you like to go ahead and mention some action items on the issue?

**Pull request closed without merging**